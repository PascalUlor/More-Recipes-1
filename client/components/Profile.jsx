/* eslint-disable */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import toastr from 'toastr';
import Spinner from 'react-md-spinner';
import NavBar from './NavBar.jsx';
import ProfileForm from './profile/ProfileForm.jsx';
import PageHeader from './profile/PageHeader.jsx';
import Footer from './Footer.jsx';
import {
  fetchProfileRequest,
  updateProfileRequest
} from '../actions/actionCreators/profileActions';
import checkImageFile from '../shared/validations/checkImageFile';
import validateInputs from '../shared/validations/profile';


/**
 * @description HOC for user profile component
 *
 * @class Profile
 *
 * @extends Component
 */
export class Profile extends Component {
  /**
   * @description creates an instance of Profile page
   *
   * @constructor
   *
   * @param { props } props - contains all user profile component properties
   *
   */
  constructor(props) {
    super(props);
    this.state = {
      fullName: '',
      username: '',
      email: '',
      location: '',
      aboutMe: '',
      initialImage: '',
      profileImage: '',
      imageFile: '',
      errors: {}
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleOnFocus = this.handleOnFocus.bind(this);
    this.handleImageChange = this.handleImageChange.bind(this);
    this.handleClick = this.handleClick.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.isValid = this.isValid.bind(this);
  }
  /**
   * @description handles fetching user profile details
   *
   * @method componentDidMount
   *
   * @returns {*} null
   */
  componentDidMount() {
    this.props.fetchProfile();
  }
  /**
   * @description receives update on lastest updates
   * @method componentWillReceiveProps
   *
   * @param {object} nextProps - object of new incoming property
   *
   * @returns {*} null
   */
  componentWillReceiveProps(nextProps) {
    let userProfile = {};
    if (typeof nextProps.updatedProfile.updatedUser !== 'undefined') {
      userProfile = nextProps.updatedProfile.updatedUser;
    } else if (typeof nextProps.fetchedProfile !== 'undefined') {
      userProfile = nextProps.fetchedProfile;
    }
    const profileImage = userProfile.profileImage === null ?
      '/images/nophoto.jpg' : userProfile.profileImage;
    this.setState({
      fullName: userProfile.fullName,
      username: userProfile.username,
      email: userProfile.email,
      location: userProfile.location === null
        ? '' : userProfile.location,
      aboutMe: userProfile.aboutMe === null
        ? '' : userProfile.aboutMe,
      profileImage,
      initialImage: profileImage
    });
  }
  /**
   * @description handles on state change
   * @method handleChange
   *
   * @param { object } event - event object containing profile details
   *
   * @returns { object } new user profile state
   */
  handleChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }
  /**
   * @description handles clicking of hidden input of type file
   * @method handleClick
   *
   * @returns {*} null
   */
  handleClick() {
    $('input[type=file]').click();
  }
  /**
   * @description handles on focus event
   * @method handleOnFocus
   * @param { object } event - event object containing profile details
   *
   * @returns { object } new user profile state
   */
  handleOnFocus(event) {
    this.setState({
      errors: Object.assign({}, this.state.errors, { [event.target.name]: '' })
    });
  }
  /**
   * @description handles on image change
   * @method handleImageChange
   *
   * @param { object } event - event object containing profile details
   *
   * @returns { object } profile image - new updated user image state
   */
  handleImageChange(event) {
    if (event.target.files && event.target.files[0]) {
      const file = event.target.files[0];
      const filereader = new FileReader();
      checkImageFile(filereader, file, (fileType) => {
        if (fileType === 'image/png' || fileType === 'image/gif'
          || fileType === 'image/jpeg') {
          this.setState({ imageFile: file });
          filereader.onload = (e) => {
            this.setState({ profileImage: e.target.result });
          };
          filereader.readAsDataURL(file);
        } else {
          toastr.remove();
          toastr.error('please provide a valid image file');
        }
      });
    } else {
      this.setState({ profileImage: this.state.initialImage, imageFile: '' });
    }
  }
  /**
   * @description handles client validation checks
   * @method isValid
   *
   * @returns { bool } true/false when form is submitted
   */
  isValid() {
    const { errors, isValid } = validateInputs(this.state);
    if (!isValid) {
      this.setState({ errors });
    }
    return isValid;
  }
  /**
   * @description handles on submit event for updating user profile
   *
   * @param { object } event - event object containing user details
   *
   * @returns { * } null
   */
  handleSubmit(event) {
    event.preventDefault();
    if (this.isValid()) {
      this.setState({ errors: {} });
      this.props.updateProfile(this.state)
        .then(() => {
          if (this.props.updateProfileError === '') {
            console.log(this.props.updatedProfile.message);
            toastr.remove();
            toastr.success(this.props.updatedProfile.message);
          } else {
            toastr.remove();
            toastr.error(this.props.updateProfileError);
          }
        });
    }
  }
  /**
   * @description renders user profile form
   *
   * @returns { jsx } jsx - renders Profile
   */
  render() {
    const { isFetching } = this.props;
    return (
      <div className="bg-faded">
        <NavBar/>
        <div className="pt-5">
          <PageHeader/>
        </div>
        <div className="container main-wrapper pt-0 mt-4">
          {isFetching ?
            <div className="text-center">
              <Spinner size={50} className="text-center mt-5"/>
            </div>
            :
            <ProfileForm
              handleChange={this.handleChange}
              handleFocus={this.handleOnFocus}
              handleClick={this.handleClick}
              handleImageChange={this.handleImageChange}
              profile={this.state}
              handleSubmit={this.handleSubmit}
              isUpdating={this.props.isProfileUpdating}/>
          }
        </div>
        <Footer/>
      </div>
    );
  }
}

Profile.propTypes = {
  isFetching: PropTypes.bool,
  fetchProfile: PropTypes.func.isRequired,
  fetchedProfile: PropTypes.shape(),
  isProfileUpdating: PropTypes.bool.isRequired,
  updatedProfile: PropTypes.shape(),
  updateProfileError: PropTypes.string,
  updateProfile: PropTypes.func.isRequired
};
/**
 * @description maps redux state to props
 *
 * @param { object } state - holds user profile state
 *
 * @return { object } props - returns mapped props from state
 */
const mapStateToProps = state => ({
  isFetching: state.userProfile.isProfileFetching,
  fetchedProfile: state.userProfile.fetchedProfile.user,
  isProfileUpdating: state.userProfile.isProfileUpdating,
  updatedProfile: state.userProfile.updatedProfile,
  updateProfileError: state.userProfile.updateProfileError,
});
/**
 * @description maps action dispatch to props
 *
 * @param { object } dispatch - holds dispatchable actions
 *
 * @return { object } props - returns mapped props from dispatch action
 */
const mapDispatchToProps = dispatch => ({
  fetchProfile: () => dispatch(fetchProfileRequest()),
  updateProfile: profileDetails =>
    dispatch(updateProfileRequest(profileDetails))
});

export default connect(mapStateToProps, mapDispatchToProps)(Profile);
