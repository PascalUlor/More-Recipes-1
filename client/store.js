import { createStore, applyMiddleware, compose } from 'redux';
import reduxImmutableStateInvariant from 'redux-immutable-state-invariant';
import thunk from 'redux-thunk';
import indexReducer from './reducers/index';


const env = process.env.NODE_ENV || development;

let middleware = compose(applyMiddleware(thunk,
    reduxImmutableStateInvariant()), window.devToolsExtension ?
  window.devToolsExtension() : f => f);

if (env === 'production') {
  middleware = applyMiddleware(thunk);
}

const store = createStore(
  indexReducer,
  middleware
);

export default store;
