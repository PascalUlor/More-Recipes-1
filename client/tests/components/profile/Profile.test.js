/* eslint-disable */
import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import toJson from 'enzyme-to-json';
import { Profile } from '../../../components/Profile.jsx';
import store from '../../../store';


const user = {
  fullName: 'Chike',
  username: 'chyke',
  email: 'chike@gmail.com',
  profileImage: null,
  location: '',
  aboutMe: '',
  errors: {},
};
const props = {
  fetchProfile: jest.fn(),
  updateProfile: jest.fn(() => Promise.resolve()),
  isProfileUpdating: false,
};
const oldState = store.getState();
const newState = {
  ...oldState,
  userProfile: {
    isProfileUpdating: false,
    updatedProfile: {},
  },
};
const newStore = mockStore(newState);
let mountWrapper = mount((
  <Provider store={newStore}>
    <Router><Profile {...props} /></Router>
  </Provider>
));
describe('Profile component', () => {
  it('should render correctly', () => {
    expect(toJson(mountWrapper)).toMatchSnapshot();
  });

  it('should call componentWillReceiveProps() for new recieved updates', () => {
    mountWrapper.find('Profile').instance().componentWillReceiveProps({
      updatedProfile: {
        updatedUser: {
          ...user,
        },
      },
    });
    expect(mountWrapper.find('Profile')
      .instance().state.username).toEqual('chyke');

    mountWrapper.find('Profile').instance().componentWillReceiveProps({
      updatedProfile: {
        updatedUser: undefined,
      },
      fetchedProfile: {
        ...user,
      },
    });
    expect(mountWrapper.find('Profile')
      .instance().state.username).toEqual('chyke');
  });

  it('should call onChange() and onClick() events', () => {
    mountWrapper.find('ProfileForm').find('TextFieldGroup')
      .at(0).find('input')
      .simulate('change', {
        target: {
          name: 'fullName',
          value: 'Chike',
        }
      });
    mountWrapper.find('ProfileForm').find('img').at(0).simulate('click');
    expect(mountWrapper.find('Profile').instance()
      .state.fullName).toEqual('Chike');
  });


  it('should call onFocus() event handler', () => {
    mountWrapper.find('ProfileForm').find('TextFieldGroup').at(0).find('input')
      .simulate('focus', {
        target: { name: 'username' },
      });
    expect(mountWrapper.find('Profile').instance().state.errors.username)
      .toEqual('');
  });

  it('should throw errors on invalidete input data', () => {
    mountWrapper.find('Profile').instance()
      .setState({
        ...user,
        fullName: '',
        username: '',
        email: '',
        location: '34 Lagos',
        aboutMe: 'ok'
      });
    mountWrapper.find('ProfileForm').find('form').simulate('submit', {
      preventDefault: jest.fn(),
    });
    expect(mountWrapper.find('Profile').instance().state.errors).toEqual({
      fullName: 'Full name is required',
      username: 'Username is required',
      email: 'Email is required',
      location: 'Provided location must contain only alphabets',
      aboutMe: 'Provided description must be at least 4 characters'
    });
  });

  it('should throw errors on invalidete input data', () => {
    mountWrapper.find('Profile').instance()
      .setState({
        ...user,
        fullName: 'Chike 45',
        username: 'ch',
        email: 'chike@gmail'
      });
    mountWrapper.find('ProfileForm').find('form').simulate('submit', {
      preventDefault: jest.fn(),
    });
    expect(mountWrapper.find('Profile').instance().state.errors).toEqual({
      fullName: 'Full name must contain only alphabets',
      username: 'Username must be at least 3 to 25 characters',
      email: 'Email is invalid'
    });
  });

  it('should throw a username error', () => {
    mountWrapper.find('Profile').instance()
      .setState({
        ...user,
        username: '2chike',
      });
    mountWrapper.find('ProfileForm').find('form').simulate('submit', {
      preventDefault: jest.fn(),
    });
    expect(mountWrapper.find('Profile').instance().state.errors).toEqual({
      username: 'Username must not start with number(s)'
    });
  });

  it('should update profile when all validations are met', () => {
    mountWrapper.find('Profile').instance().setState({ ...user });
    mountWrapper.find('ProfileForm').find('form').simulate('submit', {
      preventDefault: jest.fn(),
    });
    expect(mountWrapper.find('Profile').instance().state.errors).toEqual({});
  });
});
