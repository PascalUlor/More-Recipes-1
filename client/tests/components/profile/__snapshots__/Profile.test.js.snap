// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile component should render correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "blank",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Profile
        fetchProfile={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
          }
        }
        isProfileUpdating={false}
        updateProfile={[MockFunction]}
      >
        <div
          className="bg-faded"
        >
          <Connect(NavBar)>
            <NavBar
              authenticatedUser={
                Object {
                  "isAuthenticated": false,
                  "user": Object {},
                }
              }
              logOut={[Function]}
            >
              <div
                className="container pl-0 pr-0"
              >
                <header
                  className="mb-5"
                >
                  <nav
                    className="navbar navbar-toggleable-md navbar-light fixed-top"
                  >
                    <button
                      aria-controls="navbarItems"
                      aria-expanded="false"
                      aria-label="Toggle navigation"
                      className="navbar-toggler navbar-toggler-right"
                      data-target="#navbarItems"
                      data-toggle="collapse"
                      type="button"
                    >
                      <span
                        className="navbar-toggler-icon"
                      />
                    </button>
                    <Link
                      className="navbar-brand"
                      replace={false}
                      to="/"
                    >
                      <a
                        className="navbar-brand"
                        href="/"
                        onClick={[Function]}
                      >
                        <img
                          alt="More Recipe Logo"
                          className="d-inline-block align-center"
                          height="32"
                          src="/images/logo.png"
                          width="45"
                        />
                        <span
                          id="site-name"
                        >
                          More Recipes
                        </span>
                      </a>
                    </Link>
                    <GuestNavBar>
                      <div
                        className="collapse navbar-collapse"
                        id="navbarItems"
                      >
                        <ul
                          className="navbar-nav ml-auto text-center"
                        >
                          <li
                            className="nav-item"
                          >
                            <NavLink
                              activeClassName="active"
                              activeStyle={
                                Object {
                                  "backgroundColor": "whitesmoke",
                                  "color": "darkcyan",
                                  "fontWeight": "600",
                                }
                              }
                              ariaCurrent="true"
                              className="nav-link"
                              id="home"
                              to="/"
                            >
                              <Route
                                path="/"
                              >
                                <Link
                                  aria-current={false}
                                  className="nav-link"
                                  id="home"
                                  replace={false}
                                  to="/"
                                >
                                  <a
                                    aria-current={false}
                                    className="nav-link"
                                    href="/"
                                    id="home"
                                    onClick={[Function]}
                                  >
                                    Home
                                  </a>
                                </Link>
                              </Route>
                            </NavLink>
                          </li>
                          <li
                            className="nav-item"
                          >
                            <NavLink
                              activeClassName="active"
                              activeStyle={
                                Object {
                                  "backgroundColor": "whitesmoke",
                                  "color": "darkcyan",
                                  "fontWeight": "600",
                                }
                              }
                              ariaCurrent="true"
                              className="nav-link"
                              id="all-recipes"
                              to="/recipes"
                            >
                              <Route
                                path="/recipes"
                              >
                                <Link
                                  aria-current={false}
                                  className="nav-link"
                                  id="all-recipes"
                                  replace={false}
                                  to="/recipes"
                                >
                                  <a
                                    aria-current={false}
                                    className="nav-link"
                                    href="/recipes"
                                    id="all-recipes"
                                    onClick={[Function]}
                                  >
                                    All Recipes
                                  </a>
                                </Link>
                              </Route>
                            </NavLink>
                          </li>
                          <li
                            className="nav-item"
                          >
                            <Link
                              className="nav-link"
                              id="signin"
                              replace={false}
                              to="/signin"
                            >
                              <a
                                className="nav-link"
                                href="/signin"
                                id="signin"
                                onClick={[Function]}
                              >
                                Login
                              </a>
                            </Link>
                          </li>
                          <li
                            className="nav-item"
                          >
                            <Link
                              className="nav-link nav-button"
                              id="signup"
                              replace={false}
                              to="/signup"
                            >
                              <a
                                className="nav-link nav-button"
                                href="/signup"
                                id="signup"
                                onClick={[Function]}
                              >
                                Register
                              </a>
                            </Link>
                          </li>
                        </ul>
                      </div>
                    </GuestNavBar>
                  </nav>
                </header>
              </div>
            </NavBar>
          </Connect(NavBar)>
          <div
            className="pt-5"
          >
            <PageHeader>
              <section
                className="cover text-center"
              >
                <div
                  className="col"
                >
                  <h3
                    className="page-header p-2 mb-1"
                  >
                    My Profile
                  </h3>
                </div>
                <hr
                  id="favorites"
                />
              </section>
            </PageHeader>
          </div>
          <div
            className="container main-wrapper pt-0 mt-4"
          >
            <ProfileForm
              handleChange={[Function]}
              handleClick={[Function]}
              handleFocus={[Function]}
              handleImageChange={[Function]}
              handleSubmit={[Function]}
              isUpdating={false}
              profile={
                Object {
                  "aboutMe": "",
                  "email": "",
                  "errors": Object {},
                  "fullName": "",
                  "imageFile": "",
                  "initialImage": "",
                  "location": "",
                  "profileImage": "",
                  "username": "",
                }
              }
            >
              <form
                onSubmit={[Function]}
                role="form"
              >
                <div
                  className="row"
                >
                  <div
                    className="col-6 col-sm-3 col-md-4 col-lg-4 figure"
                  >
                    <figure
                      className="figure"
                    >
                      <img
                        alt="Site Logo"
                        className="rounded-circle img-fluid"
                        id="profile-image"
                        onClick={[Function]}
                        src=""
                      />
                    </figure>
                    <input
                      accept="image/*"
                      className="file-chooser"
                      name="imageFile"
                      onChange={[Function]}
                      type="file"
                    />
                  </div>
                  <div
                    className="col-10 col-sm-9 col-md-8 col-lg-6 profile-text"
                  >
                    <TextFieldGroup
                      font="fa fa-user"
                      label="Full Name"
                      name="fullName"
                      onChange={[Function]}
                      onFocus={[Function]}
                      placeholder="firstname lastname"
                      type="text"
                      value=""
                    >
                      <div
                        className="form-group"
                      >
                        <label
                          className="page-text"
                          htmlFor="fullName"
                        >
                          Full Name
                        </label>
                        <div
                          className="input-group mb-2 mr-sm-2 mb-sm-0"
                        >
                          <div
                            className="input-group-addon page-text"
                          >
                            <i
                              aria-hidden="true"
                              className="fa fa-user"
                            />
                          </div>
                          <input
                            className="form-control form-control-sm"
                            defaultValue=""
                            id="fullName"
                            name="fullName"
                            onChange={[Function]}
                            onFocus={[Function]}
                            placeholder="firstname lastname"
                            type="text"
                          />
                        </div>
                      </div>
                    </TextFieldGroup>
                    <TextFieldGroup
                      font="fa fa-user"
                      label="Username"
                      name="username"
                      onChange={[Function]}
                      onFocus={[Function]}
                      placeholder="enter username"
                      type="text"
                      value=""
                    >
                      <div
                        className="form-group"
                      >
                        <label
                          className="page-text"
                          htmlFor="username"
                        >
                          Username
                        </label>
                        <div
                          className="input-group mb-2 mr-sm-2 mb-sm-0"
                        >
                          <div
                            className="input-group-addon page-text"
                          >
                            <i
                              aria-hidden="true"
                              className="fa fa-user"
                            />
                          </div>
                          <input
                            className="form-control form-control-sm"
                            defaultValue=""
                            id="username"
                            name="username"
                            onChange={[Function]}
                            onFocus={[Function]}
                            placeholder="enter username"
                            type="text"
                          />
                        </div>
                      </div>
                    </TextFieldGroup>
                    <TextFieldGroup
                      font="fa fa-envelope"
                      label="Email"
                      name="email"
                      onChange={[Function]}
                      onFocus={[Function]}
                      placeholder="enter email"
                      type="text"
                      value=""
                    >
                      <div
                        className="form-group"
                      >
                        <label
                          className="page-text"
                          htmlFor="email"
                        >
                          Email
                        </label>
                        <div
                          className="input-group mb-2 mr-sm-2 mb-sm-0"
                        >
                          <div
                            className="input-group-addon page-text"
                          >
                            <i
                              aria-hidden="true"
                              className="fa fa-envelope"
                            />
                          </div>
                          <input
                            className="form-control form-control-sm"
                            defaultValue=""
                            id="email"
                            name="email"
                            onChange={[Function]}
                            onFocus={[Function]}
                            placeholder="enter email"
                            type="text"
                          />
                        </div>
                      </div>
                    </TextFieldGroup>
                    <TextFieldGroup
                      font="fa fa-map-marker"
                      label="Location"
                      name="location"
                      onChange={[Function]}
                      onFocus={[Function]}
                      placeholder="enter city where you are. eg: lagos"
                      type="text"
                      value=""
                    >
                      <div
                        className="form-group"
                      >
                        <label
                          className="page-text"
                          htmlFor="location"
                        >
                          Location
                        </label>
                        <div
                          className="input-group mb-2 mr-sm-2 mb-sm-0"
                        >
                          <div
                            className="input-group-addon page-text"
                          >
                            <i
                              aria-hidden="true"
                              className="fa fa-map-marker"
                            />
                          </div>
                          <input
                            className="form-control form-control-sm"
                            defaultValue=""
                            id="location"
                            name="location"
                            onChange={[Function]}
                            onFocus={[Function]}
                            placeholder="enter city where you are. eg: lagos"
                            type="text"
                          />
                        </div>
                      </div>
                    </TextFieldGroup>
                    <div
                      className="form-group"
                    >
                      <label
                        className="page-text"
                        htmlFor="about"
                      >
                        About Me
                      </label>
                      <textarea
                        className="form-control"
                        name="aboutMe"
                        onChange={[Function]}
                        onFocus={[Function]}
                        placeholder="write about yourself"
                        rows="5"
                        value=""
                      />
                    </div>
                    <div
                      className="text-right"
                    >
                      <button
                        className="btn btn-outline-info"
                        disabled={false}
                        type="submit"
                      >
                        Save Changes
                      </button>
                    </div>
                  </div>
                </div>
              </form>
            </ProfileForm>
          </div>
          <Footer>
            <footer
              className="footer"
            >
              <div
                className="container"
              >
                <div
                  className="row mt-4"
                >
                  <div
                    className="col-12 col-sm-8 col-md-9 col-lg-10"
                  >
                    <small
                      className="text-white"
                      id="copywrite"
                    >
                      <span
                        className="text-warning"
                      >
                        ©2018 
                      </span>
                      Bootcamp27, Andela Nigeria. All rights reserved.
                    </small>
                  </div>
                  <div
                    className="col-12 col-sm-4 col-md-3 col-lg-2 text-center"
                  >
                    <ul
                      className="list-inline"
                    >
                      <li
                        className="list-inline-item"
                      >
                        <a
                          href="#"
                        >
                          <i
                            aria-hidden="true"
                            className="fa fa-google-plus"
                          />
                        </a>
                      </li>
                      <li
                        className="list-inline-item"
                      >
                        <a
                          href="#"
                        >
                          <i
                            className="fa fa-github"
                          />
                        </a>
                      </li>
                      <li
                        className="list-inline-item"
                      >
                        <a
                          href="#"
                        >
                          <i
                            className="fa fa-facebook"
                          />
                        </a>
                      </li>
                      <li
                        className="list-inline-item"
                      >
                        <a
                          href="#"
                        >
                          <i
                            className="fa fa-twitter"
                          />
                        </a>
                      </li>
                    </ul>
                  </div>
                </div>
              </div>
            </footer>
          </Footer>
        </div>
      </Profile>
    </Router>
  </BrowserRouter>
</Provider>
`;
